packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/VirtualizedTraceView.tsx

```diff
@@ -161,17 +159,26 @@ function mergeChildrenCriticalPath(
   // Define an array to store the IDs of the span and its descendants (if the span is collapsed)
   const allRequiredSpanIds = new Set<string>([spanID]);
 
+  // Use pre-built spanMap
+  const spanMap = trace.spanMap;
+
   // If the span is collapsed, recursively find all of its descendants.
-  const findAllDescendants = (currentChildSpanIds: Set<string>) => {
-    currentChildSpanIds.forEach(eachId => {
-      const currentChildSpan = memoizedSpanByID(trace.spans).get(eachId)!;
-      if (currentChildSpan.hasChildren) {
-        currentChildSpan.childSpanIds.forEach(x => allRequiredSpanIds.add(x));
-        findAllDescendants(new Set<string>(currentChildSpan.childSpanIds));
-      }
-    });
+  const findAllDescendants = (span: Span) => {
+    if (span.hasChildren && span.childSpans.length > 0) {
+      span.childSpans.forEach(child => {
+        allRequiredSpanIds.add(child.spanID);
+        findAllDescendants(child);
+      });
+    }
   };
-  findAllDescendants(allRequiredSpanIds);
+  
+  // Start from the initially selected spans
+  allRequiredSpanIds.forEach(spanId => {
```

Comment:
this loop is redundant and causes allRequiredSpanIds to be modified while it is being iterated over, a confusing approach. Since allRequiredSpanIds always starts with a single `spanID` we can replace the loop with:
```
const startingSpan = spanMap.get(spanID);
if (startingSpan) {
  findAllDescendants(startingSpan);
}
```